src/Contexts/ContextCollection.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+
+namespace Nipwaayoni\Contexts;
+
+use Nipwaayoni\Exception\Contexts\UnsupportedContextKeyException;
+
+class ContextCollection
+{
+    private $metadata = [
+        'user' => [
+            'default' => [],
+            'mergeMethod' => 'array_replace_recursive',
+        ],
+        'custom' => [
+            'default' => [],
+            'mergeMethod' => 'array_replace_recursive',
+        ],
+        'tags' => [
+            'default' => [],
+            'mergeMethod' => 'array_replace_recursive',
+        ],
+        'env' => [
+            'default' => ['SERVER_SOFTWARE'],
+            'mergeMethod' => 'array_merge',
+        ],
+        'cookies' => [
+            'default' => [],
+            'mergeMethod' => 'array_merge',
+        ],
+    ];
+
+    private $contexts = [];
+
+    public function __construct(array $contexts)
+    {
+        $this->validateContexts($contexts);
+        $this->setContexts($contexts);
+    }
+
+    private function validateContexts(array $contexts): void
+    {
+        $unknownKeys = array_diff(array_keys($contexts), $this->allowed());
+
+        if (empty($unknownKeys)) {
+            return;
+        }
+
+        throw new UnsupportedContextKeyException(implode('|', $unknownKeys));
```

Comment:
As we have a dedicated context key exception, you could pass the keys as array and let that class format the message.

Reply:
Certainly. I admit to some short cuts for the first implementation. I've unfortunately pulled a thread that has resulted in more changes than I anticipated to make this first release viable. I'm trying to prioritize a stable public contract over internal implementations which can easily change later.

Reply:
I updated the exception as described.